tools/libxl/libxl_arm.c

```diff
@@ -786,29 +803,84 @@ static int copy_partial_fdt(libxl__gc *gc, void *fdt, void *pfdt)
+struct backlink {
+    int nodeoff;
+    int copied;
+    struct backlink *parent;
+};
 
-    for (i = 0; i < info->num_coprocs; i++) {
-        const libxl_device_coproc *coproc = &info->coprocs[i];
+static int copy_ancestor(libxl__gc *gc, void *fdt, void *pfdt,
+                         struct backlink *nodelink)
+{
+    int ret = 0;
+
+    if (nodelink && nodelink->copied == 0)
+    {
+        copy_ancestor(gc, fdt, pfdt, nodelink->parent);
+        LOG(DEBUG, "Begin node %s", fdt_get_name(pfdt, nodelink->nodeoff,
+            NULL));
+        ret = fdt_begin_node(fdt, fdt_get_name(pfdt, nodelink->nodeoff, NULL));
+        if (ret) return ret;
+        
+        ret = copy_properties(gc, fdt, pfdt, nodelink->nodeoff);
+        if (ret) return ret;
+        nodelink->copied = 1;
+    }
+    return ret;
+}
 
-        if (!coproc->path) {
-            LOG(ERROR, "Bad coproc node for the partial FDT");
-            continue;
+static int descent_for_vcoproc(libxl__gc *gc, void *fdt, void *pfdt,
+                               struct backlink *nodelink)
+{
+    int nodeoff;
+    int ret = 0;
+    struct backlink node;
+
+    LOG(DEBUG, "process node %s", fdt_get_name(pfdt, nodelink->nodeoff, NULL));
+    if ( fdt_getprop(pfdt, nodelink->nodeoff, "xen,vcoproc", NULL) )
+    {
+        LOG(DEBUG, "\"xen,vcoproc\" is found, copy ancestors");
+        ret = copy_ancestor(gc, fdt, pfdt, nodelink->parent);
+        if (ret) return ret;
+        LOG(DEBUG, "\"xen,vcoproc\" is found, copy node %s", fdt_get_name(pfdt,
+            nodelink->nodeoff, NULL));
+        ret = copy_node(gc, fdt, pfdt, nodelink->nodeoff, 0);
+    }
+    else
+    {
+        for (nodeoff = fdt_first_subnode(pfdt, nodelink->nodeoff);
+             nodeoff >= 0;
+             nodeoff = fdt_next_subnode(pfdt, nodeoff))
+        {
+            node.nodeoff = nodeoff;
+            node.parent = nodelink;
+            node.copied = 0;
+            ret = descent_for_vcoproc(gc, fdt, pfdt, &node);
+            if (ret) return ret;
         }
-
-        rc = copy_node_by_path(gc, coproc->path, fdt, pfdt);
-        if (rc < 0) {
-            LOG(ERROR, "Can't copy coproc node \"%s\" from the partial FDT", coproc->path);
-            return rc;
+        if ( nodelink->copied && nodelink->parent)
+        {
+            ret = fdt_end_node(fdt);
+            LOG(DEBUG, "Ended node %s", fdt_get_name(pfdt, nodelink->nodeoff,
+                NULL));
         }
     }
 
-    return 0;
+    return ret;
+}
+
+static int copy_vcoproc_nodes(libxl__gc *gc, void *fdt, void *pfdt,
+                              const libxl_domain_build_info *info)
+{
+    int ret = 0;
+    struct backlink node;
+
+    node.parent = NULL; /* the parent is a root subnode */
+    node.nodeoff = fdt_path_offset(pfdt, "/"); /* find the root offset */
+    node.copied = 1; /* root is already present before us */
+    ret = descent_for_vcoproc(gc, fdt, pfdt, &node);
+
+    return ret;
```

Comment:
Why don't return descent_for_vcoproc(gc, fdt, pfdt, &node);? And drop int ret then